libtenzir/builtins/functions/community_id.cpp

```diff
@@ -0,0 +1,141 @@
+public:
+  auto name() const -> std::string override {
+    return "tql2.community_id";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto args = arguments{};
+    TRY(argument_parser2::function("community_id")
+          .add(args.src_ip, "<source ip>")
+          .add(args.dst_ip, "<destination ip>")
+          .add(args.src_port, "<source port>")
+          .add(args.dst_port, "<destination port>")
+          .add(args.proto, "<transport protocol>")
+          .parse(inv, ctx));
+    return function_use::make([args = std::move(args)](evaluator eval,
+                                                       session ctx) -> series {
+      auto null_series = [&] {
+        return series::null(string_type{}, eval.length());
+      };
+      auto src_ip_series = eval(args.src_ip);
+      if (not caf::holds_alternative<ip_type>(src_ip_series.type)) {
+        diagnostic::warning("`community_id` expected `ip` as 1st argument")
+          .primary(args.src_ip)
+          .emit(ctx);
+        return null_series();
+      }
+      auto dst_ip_series = eval(args.dst_ip);
+      if (not caf::holds_alternative<ip_type>(dst_ip_series.type)) {
+        diagnostic::warning("`community_id` expected `ip` as 3rd argument")
+          .primary(args.dst_ip)
+          .emit(ctx);
+        return null_series();
+      }
+      auto src_port_series = eval(args.src_port);
+      if (not caf::holds_alternative<int64_type>(src_port_series.type)) {
+        diagnostic::warning("`community_id` expected `integer` as 2nd argument")
+          .primary(args.src_port)
+          .emit(ctx);
+        return null_series();
+      }
+      auto dst_port_series = eval(args.dst_port);
+      if (not caf::holds_alternative<int64_type>(dst_port_series.type)) {
+        diagnostic::warning("`community_id` expected `integer` as 4th argument")
+          .primary(args.dst_port)
+          .emit(ctx);
+        return null_series();
+      }
+      auto proto_series = eval(args.proto);
+      if (not caf::holds_alternative<string_type>(proto_series.type)) {
+        diagnostic::warning("`community_id` expected `string` as 5th argument")
+          .primary(args.dst_port)
+          .emit(ctx);
+        return null_series();
+      }
+      auto b = arrow::StringBuilder{};
+      check(b.Reserve(eval.length()));
+      auto src_ips = src_ip_series.as<ip_type>();
+      auto dst_ips = dst_ip_series.as<ip_type>();
+      auto src_ports = src_port_series.as<int64_type>();
+      auto dst_ports = dst_port_series.as<int64_type>();
+      auto protos = proto_series.as<string_type>();
+      for (auto i = int64_t{0}; i < eval.length(); ++i) {
+        if (src_ips->array->IsNull(i) or src_ports->array->IsNull(i)
+            or dst_ips->array->IsNull(i) or dst_ports->array->IsNull(i)
+            or protos->array->IsNull(i)) {
+          check(b.AppendNull());
+          continue;
+        }
```

Comment:
It looks like the ports are optional, at least when the protocol is not UDP or TCP. We could also add that later.

Reply:
For ICMP, the message types are used as ports. These might be difficult to get from log files. Zeek, for example, encodes them in source and destination port. But that's okay. This is something the user of this function should know. I would prefer not to default to `tcp` just to remain safe and explicit.

Reply:
I'm aware of that, but the reference implementation strongly suggests that ports are optional: 
https://github.com/corelight/pycommunityid/blob/5748e0c4808de6910abcf67b426cf8284af2da74/communityid/algo.py#L450-L457

> I would prefer not to default to tcp just to remain safe and explicit.

I fully agree with that :+1:

Reply:
> the reference implementation strongly suggests that ports are optional

Interesting, I learned something new. Thanks for being persistent!